### PR TITLE
Added swap endpoints FSM in the Flow HS Topology.

### PR DESCRIPTION
**Pull request by dpoltavets**
Related with: #3291

**MERGE NOTE: DO NOT FORGET TO CHANGE BASE BEFORE MERGING**

### CODE REVIEW

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/swapendpoints/FlowSwapEndpointsFsm.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    public void reportError(Event event) {
+        if (Event.TIMEOUT == event) {
+            reportGlobalTimeout();
+        }
+        // other errors reported inside actions and can be ignored here
+    }
+
+    @Override
+    protected String getCrudActionName() {
+        return "swap-endpoints";
```
Comment (timofei-durakov) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/swapendpoints/FlowSwapEndpointsFsm.java:96]: id remove -

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/SwapEndpointSpec.groovy [hunk 2]
```diff
@@ -820,12 +827,12 @@ switches"() {
                 new SwapFlowPayload(flow2.id, flowHelper.toFlowEndpointV2(flow2Src),
                         flowHelper.toFlowEndpointV2(flow2Dst)))
 
-        then: "An error is received (400 code)"
-        def exc = thrown(HttpClientErrorException)
-        exc.rawStatusCode == 400
-        exc.responseBodyAsString.to(MessageError).errorMessage == "Can not swap endpoints for flows: " +
-                "Failed to find path with requested bandwidth=${flow1.maximumBandwidth}: " +
-                "Switch ${flow2SwitchPair.src.dpId} doesn't have links with enough bandwidth"
+        then: "An error is received (500 code)"
```
Comment (timofei-durakov) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/SwapEndpointSpec.groovy:830]: why is it 500?

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/SwapEndpointSpec.groovy [hunk 3]
```diff
@@ -965,12 +972,12 @@ switches"() {
                 new SwapFlowPayload(flow2.id, flowHelper.toFlowEndpointV2(flow1.source),
                         flowHelper.toFlowEndpointV2(flow2.destination)))
 
-        then: "An error is received (400 code)"
-        def exc = thrown(HttpClientErrorException)
-        exc.rawStatusCode == 400
-        exc.responseBodyAsString.to(MessageError).errorMessage == "Can not swap endpoints for flows: " +
-                "Failed to find path with requested bandwidth=${flow2.maximumBandwidth}: " +
-                "Switch ${flow1SwitchPair.src.dpId} doesn't have links with enough bandwidth"
+        then: "An error is received (500 code)"
```
Comment (timofei-durakov) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/SwapEndpointSpec.groovy:975]: the same

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/SwapEndpointSpec.groovy [hunk 4]
```diff
@@ -1137,6 +1144,76 @@ switches"() {
...
+
+        when: "Try to swap flow src endoints, but flow1 src switch does not respond"
+        def blockData = switchHelper.knockoutSwitch(swPair1.src, mgmtFlManager)
+        database.setSwitchStatus(swPair1.src.dpId, SwitchStatus.ACTIVE)
+        northbound.swapFlowEndpoint(new SwapFlowPayload(flow1.flowId, flow2.source, flow1.destination),
+                new SwapFlowPayload(flow2.flowId, flow1.source, flow2.destination))
+
+        then: "Receive error response"
+        def exc = thrown(HttpServerErrorException)
+        exc.rawStatusCode == 500
```
Comment (timofei-durakov) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/SwapEndpointSpec.groovy:1174]: and here too

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/swapendpoints/actions/RevertUpdateRequestAction.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        stateMachine.setFlowResponses(new ArrayList<>());
+
+        Flow firstOriginalFlow = stateMachine.getFirstOriginalFlow();
+        Flow secondOriginalFlow = stateMachine.getSecondOriginalFlow();
+
+        sendRevertUpdateCommand(firstOriginalFlow, secondOriginalFlow.getFlowId(), stateMachine);
+        sendRevertUpdateCommand(secondOriginalFlow, firstOriginalFlow.getFlowId(), stateMachine);
+    }
+
+    private void sendRevertUpdateCommand(Flow flow, String anotherFlowId, FlowSwapEndpointsFsm stateMachine) {
```
Comment (sergii-iakovenko) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/swapendpoints/actions/RevertUpdateRequestAction.java:50]: What if one of the flows has "in-progress" status (like network changes trigger a reroute)? Will this leave the changes not reverted?

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/swapendpoints/actions/OnFinishedAction.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    public OnFinishedAction(PersistenceManager persistenceManager, FlowOperationsDashboardLogger dashboardLogger) {
+        super(persistenceManager, dashboardLogger);
+    }
+
+    @Override
+    protected Optional<Message> performWithResponse(State from, State to, Event event, FlowSwapEndpointsContext context,
+                                                    FlowSwapEndpointsFsm stateMachine) {
+        log.info("Swap endpoints operation completed successfully.");
+
+        updateFlowStatus(stateMachine.getFirstFlowId(), stateMachine);
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/swapendpoints/actions/OnFinishedAction.java]: Yes, it should be. Fixed.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/swapendpoints/actions/OnFinishedWithErrorAction.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        }
+
+        if (!Event.VALIDATION_ERROR.equals(event)) {
+            saveActionToHistory(stateMachine,
+                    stateMachine.getFirstFlowId(), stateMachine.getSecondFlowId(), data.getErrorDescription());
+            saveActionToHistory(stateMachine,
+                    stateMachine.getSecondFlowId(), stateMachine.getFirstFlowId(), data.getErrorDescription());
+        }
+
+        updateFlowStatus(stateMachine.getFirstFlowId(), stateMachine);
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/swapendpoints/actions/OnFinishedWithErrorAction.java]: Yes, it should be. Fixed.

**Pull request merged**